crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -351,24 +351,43 @@ where
 
     /// Create a new package
     pub fn new_package<'p>(
-        &self,
+        &mut self,
         modules: Vec<move_binary_format::CompiledModule>,
         dependencies: impl IntoIterator<Item = &'p MovePackage>,
         version: Option<SequenceNumber>,
-    ) -> Result<Object, ExecutionError> {
+    ) -> Result<ObjectID, ExecutionError> {
         // wrap the modules in an object, write it to the store
-        Object::new_package(
+        let object = Object::new_package(
             modules,
             version.unwrap_or(OBJECT_START_VERSION),
             self.tx_context.digest(),
             self.protocol_config.max_move_package_size(),
             dependencies,
-        )
+        )?;
+        let object_id = object.id();
+        self.new_packages.push(object);
+        Ok(object_id)
     }
 
-    /// Add the package to the new package published
-    pub fn add_package(&mut self, package_obj: Object) {
-        self.new_packages.push(package_obj)
+    /// Create a package upgrade from `previous_package` with `new_modules` and `dependencies`
+    pub fn upgrade_package<'p>(
+        &mut self,
+        previous_package: &MovePackage,
+        new_modules: Vec<move_binary_format::CompiledModule>,
+        dependencies: impl IntoIterator<Item = &'p MovePackage>,
+    ) -> Result<ObjectID, ExecutionError> {
+        let new_package_object_id = self.tx_context.fresh_id();
```

Comment:
Note that here we are creating a new object ID for the upgraded package and pushing it down into the object creation. 

This is fine since the object ID of a package object grabs the ID from the field on package (which we're populating with the fresh ID that we're creating here).